### PR TITLE
Implement Proof of Work in Client

### PR DESCRIPTION
**Pull request by agusaldasoro**
Configure `LambdaClient` and `ContentClient` to execute Proof of Work to obtain the JWT that will be used as Cookie in all requests. Also, configure the setImmediate that will check the header "Set-Cookie" for every request made to get a new JWT in case it expires.

To enable this flow, the environment variable `PROOF_OF_WORK` must have its value as `true`.

### Breaking Change
Params of Client moved into an options object.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -45,7 +45,8 @@
   "dependencies": {
     "@types/form-data": "^2.5.0",
     "async-iterator-to-array": "0.0.1",
-    "dcl-catalyst-commons": "^5.0.0",
+    "cookie": "^0.4.1",
+    "dcl-catalyst-commons": "tag-add-cookies-v7",
```
Comment (menduz) [inline package.json:49]: Should we publish a fixed version before merging this PR?

File: src/ContentClient.ts [hunk 2]
```diff
@@ -49,12 +51,19 @@ export class ContentClient implements ContentAPI {
...
+    this.deploymentBuilderClass = deploymentBuilderClass ?? DeploymentBuilder
     this.fetcher =
       fetcher ??
       new Fetcher({
         headers: getHeadersWithUserAgent('content-client')
       })
-    this.deploymentBuilderClass = deploymentBuilderClass ?? DeploymentBuilder
+
+    if (PROOF_OF_WORK) {
+      setImmediate(async () => {
```
Comment (menduz) [inline src/ContentClient.ts:62]: - I'd like to understand the reasoning behind `setImmediate` vs simply `await`
- `setImmediate` does not handle exceptions. If the async function fails it would cascade to the event loop and close in future versions and trigger a `unhandledRejection` in the process
- `setImmediate` is not available in all JS engines, i.e. it is not present in firefox

File: src/LambdasClient.ts [hunk 3]
```diff
@@ -27,6 +29,13 @@ export class LambdasClient implements LambdasAPI {
       new Fetcher({
         headers: getHeadersWithUserAgent('lambdas-client')
       })
+
+    if (PROOF_OF_WORK) {
+      setImmediate(async () => {
+        const powAuthBaseUrl = new URL(this.lambdasUrl).origin
+        await setJWTAsCookie(this.fetcher, powAuthBaseUrl)
+      })
```
Comment (menduz) [inline src/LambdasClient.ts:37]: Same as above

File: src/utils/ProofOfWork.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+// Code copied from https://github.com/decentraland/pow-authorization-server
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const crypto = require('crypto')
```
Comment (menduz) [inline src/utils/ProofOfWork.ts:4]: I think that if it is possible, we should use `import {createHash,randomBytes} from 'crypto'` to add type safety

File: src/ports/Jwt.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+export async function setJWTAsCookie(fetcher: Fetcher, baseUrl: string): Promise<void> {
+  const jwt = await obtainJWTWithRetry(fetcher, baseUrl, 3)
+
+  fetcher.overrideDefaults({ cookies: { JWT: jwt } })
+  fetcher.overrideSetImmediate(async (response: Response) => {
+    if (removedJWTCookie(response)) {
+      const jwt = await obtainJWT(fetcher, baseUrl)
+      if (!!jwt) {
+        fetcher.overrideDefaults({ cookies: { JWT: jwt } })
```
Comment (menduz) [inline src/ports/Jwt.ts:53]: Does setting explicit cookies work in the browser?

File: src/ports/Jwt.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  } catch {
+    return false
+  }
+}
+
+export async function setJWTAsCookie(fetcher: Fetcher, baseUrl: string): Promise<void> {
+  const jwt = await obtainJWTWithRetry(fetcher, baseUrl, 3)
+
+  fetcher.overrideDefaults({ cookies: { JWT: jwt } })
+  fetcher.overrideSetImmediate(async (response: Response) => {
```
Comment (menduz) [inline src/ports/Jwt.ts:49]: I want to make sure that the `setImmediate` thing we are planning is intended to run \*after* the request connection is open. And that it also is used \*only* for JWT, because it seems like a generic name, that can be overwritten by other custom behavior in the code.

File: package.json (hunk 1 above)
Comment (jmoguilevsky) [inline package.json:49]: this will be a release version, right?

**Review by menduz — CHANGES_REQUESTED** [review]
After I read through the rest of the code, I'm seeing a risk on the `setImmediate` behaviors and implementation. The whole logic has been very hard to follow, and it generated questions like 
> Does the implementation of JWT breaks if anyone else calls `fetcher.overrideSetImmediate`?

> Is this intended to happen right after the request? Or is it intended to modify the response before return?

This is probably not the best PR to perform the change I am about to propose, I'd like to open the conversation to it: I want to present the idea of middleware or transformer functions instead of `overrideSetImmediate`.

```ts
class Fetcher {
  responseMiddleware: (r: Response) => Promise<Response> = async res => res
  requestMiddleware: (r: Request) => Promise<Request> = async req => req
  
  doRequest(req: Request){
    const req = await requestMiddleware(req)
    const res = await fetch(req)
    return responseMiddleware(res)
  }
}


function addJwtToFetcher(fetcher) {
     const nextRequest = fetcher.requestMiddleware
     fetcher.requestMiddleware = async (request) => {
        request.headers.set('Cookie', await getJwt())
        return nextRequest(request)
     }

     const nextResponse = fetcher.responseMiddleware
     fetcher.responseMiddleware = async (response) => {
        if(response.headers.has('set-cookie')){
          // do something with cookie
        }
        return nextResponse(response)
     }
}
```

I think the approach is better because setImmediate is not present in every environment and it would schedule the event to the `thread.yield`, that provokes undesired side effects and race conditions which are very hard to debug.

**Review by jmoguilevsky — APPROVED** [review]

File: package.json (hunk 1 above)
Comment (agusaldasoro) [reply package.json:49]: Yes! This is blocked by https://github.com/decentraland/catalyst-commons/pull/68

File: src/ContentClient.ts (hunk 2 above)
Comment (agusaldasoro) [reply src/ContentClient.ts:62]: The rationale behind using `setImmediate` was to avoid blocking the request with this logic, but after discussing this with @jmoguilevsky too I'm removing it and simply using an `await`.

File: src/ports/Jwt.ts (hunk 5 above)
Comment (agusaldasoro) [reply src/ports/Jwt.ts:53]: I'm not sure, anyway, the main user of the Catalyst Client is not the browser, each Catalyst is.

File: src/ports/Jwt.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        }
+      }
+      return request
+    },
+    responseMiddleware: async (response: Response) => {
+      if (removedJWTCookie(response)) {
+        const jwt = await obtainJWT(fetcher, baseUrl)
+        if (!!jwt) {
+          fetcher.overrideDefaults({ cookies: { JWT: jwt } })
+        }
```
Comment (menduz) [inline src/ports/Jwt.ts:83]: I think the most correct implementation of the response should be only to remove the JWT, so it can be regenerated next time in `requestMiddleware` line 71
```suggestion
        fetcher.overrideDefaults({ cookies: { } })
```

File: src/ports/Jwt.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+export function removedJWTCookie(response: Response): boolean {
+  try {
+    const headers = response.headers
+    if (headers) {
+      const setCookie = headers.get('Set-Cookie')
+      if (setCookie && setCookie.includes('JWT=')) {
+        const cookies = cookie.parse(setCookie)
+        return cookies.JWT == ''
```
Comment (nchamo) [inline src/ports/Jwt.ts:38]:
```suggestion
        return cookies.JWT === ''
```

File: src/ports/Jwt.ts [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+      LOGGER.warn('[POW] Could not get a JWT from Pow Auth Server.')
+    }
+    return jwtResponse.jwt
+  } catch (error) {
+    LOGGER.warn(`[POW] Could not get a JWT from Pow Auth Server, due to: ${error}`)
+    return ''
+  }
+}
+
+export function removedJWTCookie(response: Response): boolean {
```
Comment (nchamo) [inline src/ports/Jwt.ts:31]: I think `isJWTCookiePresent` or `isJWTCookieRemoved` would be a clearer name. I initially thought this function removed the cookie

File: src/ports/Jwt.ts [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        return cookies.JWT == ''
+      }
+    }
+    return false
+  } catch {
+    return false
+  }
+}
+
+export function noJWTinCookie(request: CrossFetchRequest): boolean {
```
Comment (nchamo) [inline src/ports/Jwt.ts:47]: We have too many functions that check if the cookie is present haha. Could we make their names clearer so that it is easier to understand the differences?

File: src/utils/catalystList.ts [hunk 11]
```diff
@@ -16,24 +16,25 @@ export async function getApprovedListFromContract(network: 'mainnet' | 'ropsten'
...
-export async function getUpdatedApprovedListWithoutQueryingContract({
-  preKnownServers,
-  origin,
-  requiredLists,
-  fetchApprovedCatalysts
-}: {
+export type KnownServersOptions = {
   preKnownServers: { list: { address: string }[] } | { network: 'mainnet' | 'ropsten' }
   origin: string
+  proofOfWorkEnabled: boolean
```
Comment (nchamo) [inline src/utils/catalystList.ts:22]: Optional

File: src/ports/Jwt.ts [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    return true
+  }
+}
+
+function hasJWTCookie(cookieValue: string): boolean {
+  const cookies = cookie.parse(cookieValue)
+  return cookies.JWT ?? '' != ''
+}
+
+export function setJWTAsCookie(fetcher: Fetcher, baseUrl: string): void {
```
Comment (nchamo) [inline src/ports/Jwt.ts:70]: Correct me if I'm wrong, but this function is no longer setting the jwt. It just injects the middleware, right? If so, then let's change the function name

**Review by nchamo — COMMENTED** [review]
Great work, added some comments

File: src/ports/Jwt.ts [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+export function configureJWTMiddlewares(fetcher: Fetcher, baseUrl: string): void {
+  let lastFailedPowEndpointTimestamp: number = 0
+  let minutesToAdd: number = ms('5m')
+  let isRequestingJWT: boolean = false
+  fetcher.overrideDefaults({
+    requestMiddleware: async (request: CrossFetchRequest) => {
+      if (missingJWTInRequest(request) && !isRequestingJWT) {
+        if (lastFailedPowEndpointTimestamp + minutesToAdd < Date.now()) {
+          isRequestingJWT = true
```
Comment (nchamo) [inline src/ports/Jwt.ts:79]: I would add a try/catch and set `isRequestingJWT` to false on the finally. Because if `obtainJWT` throws an exception, I'm not sure how everything will end up being

File: test/ports/Jwt.spec.ts [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+    expect(body.complexity).toEqual(4)
+    expect(body.nonce).toEqual('aNonce')
+  })
+
+  it('should return the jwt as string', async () => {
+    expect(jwt).toEqual('aJWT')
+  })
+})
+
+describe('removedJWTCookie', () => {
```
Comment (nchamo) [inline test/ports/Jwt.spec.ts:58]: Since you updated the function names, could you also update the test titles?

**Review by nchamo — APPROVED** [review]
Great work!

File: src/ports/Jwt.ts [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+import log4js from 'log4js'
+import ms from 'ms'
+import { generateNonceForChallenge } from '../utils/ProofOfWork'
+
+const LOGGER = log4js.getLogger('JWTPort')
+
+export async function obtainJWT(fetcher: Fetcher, catalystUrl: string): Promise<string | undefined> {
+  try {
+    const response = await fetcher.fetchJson(catalystUrl + '/pow-auth/challenge')
+    const body = JSON.parse(JSON.stringify(response))
```
Comment (menduz) [inline src/ports/Jwt.ts:12]: Is this a cheap `clone(response)`?

File: src/ports/Jwt.ts (hunk 13 above)
Comment (menduz) [reply src/ports/Jwt.ts:79]: +1

File: src/ports/Jwt.ts [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+            const jwt = await obtainJWT(fetcher, baseUrl)
+            if (!!jwt) {
+              fetcher.overrideDefaults({ cookies: { JWT: jwt } })
+              lastFailedPowEndpointTimestamp = 0
+              minutesToAdd = ms('5m')
+            } else {
+              lastFailedPowEndpointTimestamp = Date.now()
+              minutesToAdd = 2 * minutesToAdd
+            }
+          } catch {
```
Comment (menduz) [inline src/ports/Jwt.ts:90]: If not handling the exception, either fail or handle/print it somewhere
```suggestion
```

**Review by menduz — APPROVED** [review]
Small comments, otherwise looking good

File: src/ports/Jwt.ts (hunk 15 above)
Comment (agusaldasoro) [reply src/ports/Jwt.ts:12]: Oops, that line shouldn't be necessary. Nice catch!
